torch/csrc/jit/passes/onnx/peephole.cpp

```diff
@@ -830,6 +830,153 @@ void removeMaxPoolUnusedOutput(Block* b) {
   }
 }
 
+// recursive function that deletes all nodes between NegativeLogLikelihoodLoss
+// and LogSoftmax
+void recursive_del(Node* node) {
+  if (node->kind() == onnx::LogSoftmax) {
+    return;
+  }
+
+  std::vector<Node*> nodeList;
+  for (size_t i = 0; i < node->inputs().size(); ++i) {
+    nodeList.push_back(node->input(i)->node());
+  }
+  node->removeAllInputs();
+
+  if (node->kind() != onnx::NegativeLogLikelihoodLoss) {
+    bool del = true;
+    for (size_t i = 0; i < node->outputs().size(); ++i) {
+      if (!node->outputs()[i]->uses().empty()) {
+        del = false;
+        break;
+      }
+    }
+    if (del) {
+      node->destroy();
+    }
+  }
+
+  for (Node* n: nodeList) {
+    recursive_del(n);
+  }
+}
+
+Node* createSoftmaxCrossEntropyNode(Block* b, graph_node_list_iterator it, std::vector<Value*> values) {
+  Node* softmaxCrossEntropyNode = b->owningGraph()->create(onnx::SoftmaxCrossEntropyLoss, it->outputs().size());
+  for (size_t i = 0; i < softmaxCrossEntropyNode->outputs().size(); ++i) {
+    softmaxCrossEntropyNode->outputs()[i]->copyMetadata(it->outputs()[i]);
+  }
+  for (Value* value : values) {
+    softmaxCrossEntropyNode->addInput(value);
+  }
+  return softmaxCrossEntropyNode;
+}
+
+// This optimization fuses LogSoftmax and NegativeLogLikelihoodLoss operators into
+// one operator: SoftmaxCrossEntropyLoss.
+static void fuseLogSoftmaxNllLoss(Block* b) {
+  for (auto it = b->nodes().begin(), end = b->nodes().end(); it != end; ++it) {
+    for (auto* child_block : it->blocks()) {
+      fuseLogSoftmaxNllLoss(child_block);
+    }
+    if (it->kind() == onnx::NegativeLogLikelihoodLoss) {
+      auto prev = it->input(0)->node();
+      Node* origLogSoftmaxNode;
+      Node* origNllLossNode;
+      std::vector<Node*> deleteNodes;
+      Node* softmaxCrossEntropyNode = nullptr;
+      // if the input is 2D
+      if (prev->kind() == onnx::LogSoftmax) {
+        origLogSoftmaxNode= it->input(0)->node();
+        origNllLossNode = *it;
+        deleteNodes.push_back(origLogSoftmaxNode);
+      // if the input is 4D
+      } else if (prev->kind() == onnx::Transpose &&
+                 prev->input(0)->node()->kind() == onnx::LogSoftmax) {
+        auto logSoftmaxNode = prev->input(0)->node();
+        origLogSoftmaxNode = logSoftmaxNode->input(0)->node();
+        origNllLossNode = *it;
+        prev->removeAllInputs();
```

Comment:
We need to remove all the inputs of the node so that input nodes can be deleted. Before deletion there is a check node->outputs[I].uses().empty()

Reply:
I think Bowen's point is that once you delete all the uses (connections) of nodes in the subgraph in between the LogSoftmax and NLLLoss, the subgraph becomes orphan (not used), and will be deleted as part of "dce" pass. 
This is to avoid explicitly:
        deleteNodes.push_back(origLogSoftmaxNode);
        deleteNodes.push_back(logSoftmaxNode);
        deleteNodes.push_back(prev);

Reply:
I see, in that case I don't need function recursion_del